lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs

```diff
@@ -98,28 +98,48 @@ public void SetException(Exception e)
         {
             lock (_heartBeatManager)
             {
-                try
+                if (_lastException.IsPresent())
                 {
-                    if (!_lastException.IsPresent())
-                    {
-                        _lastException = Optional<Exception>.Of(e);
-                    }
+                    return;
+                }
 
-                    State = TaskState.Failed;
+                _lastException = Optional<Exception>.Of(e);
+                State = TaskState.Failed;
+
+                try
+                {
                     _taskLifeCycle.Stop();
                 }
-                finally
+                catch (Exception stopException)
                 {
-                    Heartbeat();
+                    Utilities.Diagnostics.Exceptions.Caught(
+                        stopException, Level.Error, "Exception thrown on Task Stop", LOGGER);
                 }
+
+                Heartbeat();
             }
         }
 
         public void SetResult(byte[] result)
         {
             lock (_heartBeatManager)
             {
+                if (_state == TaskState.Failed)
+                {
+                    if (!_lastException.IsPresent())
+                    {
+                        throw new ApplicationException("Expected failed Task to already have set an Exception.");
```

Comment:
right, can we crash evaluator in the SetException right away?
Basically we want to guarantee state integrity, it is easier to do during state change, otherwise we may need to do these checks in multiply places afterwards.


---
In reply to: [67907809](https://github.com/apache/reef/pull/1042#discussion_r67907809) [](ancestors = 67907809)